File: src/content/docs/topics/cs/coding-problems/arrays-and-hashing/454-4sum-ii.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

function fourSumCount(nums1: number[], nums2: number[], nums3: number[], nums4: number[]): number {
    // TODO: implement
    return 0;
}

function _runTests(): void {
    assert(fourSumCount([1,2], [-2,-1], [-1,2], [0,2]) === 2);
    assert(fourSumCount([0], [0], [0], [0]) === 1);
    assert(fourSumCount([-1,-1], [-1,1], [-1,1], [1,-1]) === 6);
    // perf
    const n = 500;
    const arr = Array.from({length: n}, (_, i) => i);
    const t0 = performance.now();
    fourSumCount(arr, arr, arr, arr);
    console.log(`perf fourSumCount n=${n}: ${(performance.now() - t0).toFixed(1)}ms`);
    console.log('all tests pass');
}

_runTests();

File: src/content/docs/topics/cs/coding-problems/arrays-and-hashing/454-4sum-ii.go
package main

import "fmt"

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func fourSumCount(nums1 []int, nums2 []int, nums3 []int, nums4 []int) int {
	// TODO: implement
	return 0
}

func runTests() {
	assert(fourSumCount([]int{1, 2}, []int{-2, -1}, []int{-1, 2}, []int{0, 2}) == 2, "test 1")
	assert(fourSumCount([]int{0}, []int{0}, []int{0}, []int{0}) == 1, "test 2")
	assert(fourSumCount([]int{-1, -1}, []int{-1, 1}, []int{-1, 1}, []int{1, -1}) == 6, "test 3")
	fmt.Println("all tests pass")
}

func main() { runTests() }
